### DISCUSSION TITLE
Parsing structure that depends on base offset of struct?

### DISCUSSION THREAD

**Discussion by landaire** [jam1garner/binrw#203] [Q&A] [answered]
I've never used `binrw` or `binread` before, so I'm looking for some suggestions on how to best implement this scenario. I have some file format that does lookups based upon the base offset of the struct. Here's my 010editor template:

```c
typedef struct {
    local int64 base_off = FTell();

    uint64 resource_ptr;
    uint64 filename_ptr;

    local int64 off = FTell();
    FSeek(base_off + (int64)filename_ptr);
    string filename;
    FSeek(off);

    uint64 id;
    uint64 parent_id;
} Block;
```

For example, if the current offset is `0x10` and `filename_ptr` is `0x100`, the null-terminated `filename` will be located at `0x110`.

Here is how I'm currently able to parse this: https://gist.github.com/landaire/c8cbd286c21f527a11ddc67b314877f3

But this doesn't feel like it's in spirit with the declarative nature of the crate.

**Comment by kitlith** [comment, marked as answer]
You can use the PosValue helper struct to get the position. `PosValue<()>` if you just want the position and don't want to wrap a value.

**Reply by landaire** [reply]
Thank you! This did the trick:

```rust
#[derive(Debug, BinRead)]
struct Volume {
    this_offset: PosValue<()>,
    len: u64,
    name_ptr: u64,
    volume_id: u64,
    #[br(seek_before = SeekFrom::Start(this_offset.pos+ name_ptr), restore_position)]
    filename: NullString,
}
```

**Reply by mon** [reply]
Is there an equivalent construct for writes?

**Reply by csnover** [reply]
Hi @mon. Can you open a new discussion question and include some sample of what you are trying to do? Writing offsets is more complicated so there is more than one possible answer depending on the context.

**Reply by mon** [reply]
Sure! For any future viewers, I opened it here: https://github.com/jam1garner/binrw/discussions/229
